### PR TITLE
librbd: potential race when replaying journal ops

### PR DESCRIPTION
**Pull request by dillaman**
Fixes: http://tracker.ceph.com/issues/16198
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

File: src/librbd/journal/Replay.cc [hunk 1]
```diff
@@ -802,8 +799,15 @@ void Replay<I>::handle_op_complete(uint64_t op_tid, int r) {
...
-    on_flush->complete(0);
+
+  // shut down request might have occurred while lock was
+  // dropped -- handle if pending
+  Mutex::Locker locker(m_lock);
+  assert(m_in_flight_op_events > 0);
+  --m_in_flight_op_events;
+  if (m_flush_ctx != nullptr && m_in_flight_op_events == 0 &&
+      (m_in_flight_aio_flush + m_in_flight_aio_modify) == 0) {
+    m_image_ctx.op_work_queue->queue(m_flush_ctx, 0);
```
Comment (trociny) [inline src/librbd/journal/Replay.cc]: @dillaman m_flush_ctx can be completed from both handle_aio_flush_complete and handle_op_complete. I wonder if a race is possible here and it is completed twice?

Comment (dillaman) [inline src/librbd/journal/Replay.cc]: New IO and op events shouldn't be processed after ```shut_down``` is invoked. With the IO and op counters, it should only be able to be invoked one time -- depending on whether an IO or op finished last.

Comment (trociny) [inline src/librbd/journal/Replay.cc]: Ah, I see, makes sense.

**Comment by trociny** [comment]
lgtm

**Pull request merged**